atomics/src/Lock_Array_SYCL.cpp

```diff
@@ -63,7 +53,10 @@ void finalize_lock_arrays_sycl<int>(sycl::queue q) {
   sycl::free(SYCL_SPACE_ATOMIC_LOCKS_NODE_h, q);
   SYCL_SPACE_ATOMIC_LOCKS_DEVICE_h = nullptr;
   SYCL_SPACE_ATOMIC_LOCKS_NODE_h = nullptr;
+#ifdef DESUL_ATOMICS_ENABLE_SYCL_SEPARABLE_COMPILATION
```

Comment:
I see also guarded it for CUDA and HIP but it is not clear to me why we did.
Not asking you to change but do you know why that would be needed?

Reply:
My best guess is that we do it for the same reason as we assign `nullptr` to `[CUDA|HIP]_SPACE_ATOMIC_LOCKS_[DEVICE|NODE]_h` and that that is to avoid dangling pointers to freed allocations in case we don't safeguard probably against executing kernels after this function has been called. This is very similar to all in the `nullptr` assignments in `ExecutionSpace::finalize`.

Reply:
Yeah, that's interesting.

With RDC, we have only one device variable that is the same for all translation units. Assuming that, `init_lock_arrays_*` was called, `copy_*_lock_arrays_to_device` won't do anything since the copy is only executed once. Otherwise, we actually copy the `nullptr`.
Without RDC, we have one device variable per translation unit that we are setting up via `ensure_*_lock_arrays_on_device` which basically calls `copy_*_lock_arrays_to_device` again. In this case, we can't reset all the device variables anyway.

I think it makes the most sense to either remove the guard here or add it in `init_lock_arrays_*` as well. If we assume that `init_lock_arrays_*` is called before `finalize_lock_arrays_*` we can also remove the call here. I'm happy to open a follow-up pull request that discusses/addresses that uniformly across the backends.

Reply:
That sounds like a reasonable approach, but I think in practice we can end up in a situation with more than one lock array _even with RDC turned on_ because of the split between separate dynamic libraries.  I sincerely hope nobody's doing that, but we're only guaranteed to have a single array _per dynamic library_ not globally.

Reply:
Let's close this parenthesis and address this in a follow up.
In the meantime I think this is ready.

Reply:
I agree, looks really good actually.